### ISSUE TITLE
Ensure full commit timing out on replication of large database

### ISSUE THREAD

**Issue by jdai1** [apache/couchdb#4639]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description
I am trying to replicate the npm registry through couchdb. However, I keep getting the following errors:

{"error":"checkpoint_commit_failure","reason":"Failure on source commit: {'EXIT',{http_request_failed,\"POST\",\n                             \"https://skimdb.npmjs.com/registry/_ensure_full_commit\",\n                             {error,{error,req_timedout}}}}"}

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT give us access or passwords to your actual CouchDB! )

## Steps to Reproduce

I am using version 3.3.2. When post a request to the _ensure_full_commit endpoint, it returns something that looks like this:

{"ok":true,"instance_start_time":"1686442464"}

However, based on the documentation, it should be returning an instance start time of 0.

[NOTE]: # ( Include commands to reproduce, if possible. curl is preferred. )

## Expected Behaviour

[NOTE]: # ( Tell us what you expected to happen. )

## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )
[TIP]:  # ( You can paste the output of curl http://YOUR-COUCHDB:5984/ here. )

{"couchdb":"Welcome","version":"3.3.2","git_sha":"11a234070","uuid":"358bddb8b0671d6dc8774db0a626edef","features":["access-ready","partitioned","pluggable-storage-engines","reshard","scheduler"],"vendor":{"name":"The Apache Software Foundation"}}

* CouchDB version used:
* Browser name and version:
* Operating system and version:

## Additional Context

[TIP]:  # ( Add any other context about the problem here. )

**Comment by nickva** [comment]
As `{error,req_timedout}` indicates a timeout.  Try to increase the replication connection timeout.

https://docs.couchdb.org/en/stable/config/replicator.html#replicator/connection_timeout

The `instance_start_time` change, to return a db creation timestamp instead of `0` is a 3.3 change . We did mention it in the [release notes](https://docs.couchdb.org/en/stable/whatsnew/3.3.html#bugfixes) but did not update the documentation. https://github.com/apache/couchdb/issues/3901

In all of the 3.x versions, `_ensure_full_commit` doesn't actually do anything, it's a compatibility no-op. All the documents updates are now committed during doc update operation. The only thing the API endpoint returns is the database instance start time value. So, it is a bit strange that `_ensure_full_commit` is timing out, unless the endpoint is not a 3.x Apache CouchDB version but an older version or another couch implementation.

**Comment by jdai1** [comment]
Hi Nick! Thank you so much for your help. I believe the source database of the replication is an older version of CouchDB. I will definitely try increasing the connection timeout. Another quick question, if I change the replication config in the local.ini file, will it apply the config if I send a new replication HTTP request, even if the new replication task is a continuation of the existing replication?

**Comment by jdai1** [comment]
Also, on my active_tasks page, it is still saying I have many bulk_get_attempts, even though my bulk_get_attempts is set to false in my configuration file.

<img width="569" alt="Screenshot 2023-06-12 at 1 20 43 PM" src="https://github.com/apache/couchdb/assets/44037521/9b8607d6-744e-4626-9723-13a767f0a1c2">
<img width="853" alt="Screenshot 2023-06-12 at 1 21 33 PM" src="https://github.com/apache/couchdb/assets/44037521/624d1ec6-f5c0-47a7-bb0b-6e47119b4906">

**Comment by jdai1** [comment]
Oh and finally, do you have any suggestions as to how I can replicate from the outdated couchdb instance when faced with the timeout request on the _ensure_full_commit endpoint? Thank you so much!

**Comment by nickva** [comment]
> Oh and finally, do you have any suggestions as to how I can replicate from the outdated couchdb instance when faced with the timeout request on the _ensure_full_commit endpoint? Thank you so much!

You can try disable bulk_get usage, increase the timeout, and reduce the number of workers a bit. 200 concurrent workers may be a bit too much for the endpoint to handle? Try reducing their number to say 10 and see what it does. Try reducing your batch size as well a bit.

Replicating in general implies being able to update the _local checkpoint document on the source endpoint. If the source endpoint doesn't allow that, depending how that's implemented it may manifest as a timeout (if they just close the connection and don't respond). In that case you can try disable checkpointing altogether with: https://docs.couchdb.org/en/stable/config/replicator.html#replicator/use_checkpoints As the option implies this would disable checkpointing so on failure, it will retry to scan all the docs again next time it tries to run.

**Comment by jdai1** [comment]
Sorry to keep bothering you, would you happen to know why I also get these errors?
```
{"error":"error","reason":"max_backoff"}
```

**Comment by nickva** [comment]
@jdai1 no problem at all, thanks for reaching out. 

The `max_backoff` error indicates that some replicator API request was throttled too much and it exceeded a max retry interval before the job crashes and restarts.

Those errors are usually emitted if a [429 error code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429) is returned from an HTTP API endpoint. Replication jobs would retry individual requests with an exponential backoff, and in the case of 429 errors it will use the [AIMD](https://en.wikipedia.org/wiki/Additive_increase/multiplicative_decrease) - additive increase, multiplicative decrease algorithm, similar to TCP to try to match the throttling rate. But, there is a limit to those failures as well, and if those requests keep getting throttled, it will crash the whole job, and restart it with a simpler configuration. It will apply these setting to those jobs:

```
[
        {checkpoint_interval, 5000},
        {worker_processes, 2},
        {worker_batch_size, 100},
        {http_connections, 5}
 ]
```

and re-schedule the job to run again.

Even if the replication jobs will keep crashing (they are in the `crashing` state), they should still be restarting from time to time, trying as hard as they can to make some progress.

So, to summarize, individual replication API requests are retried on failure, and then the whole replication job is continuously restarted as well. See the advanced replication [guide](https://docs.couchdb.org/en/stable/replication/replicator.html#replication-states) for more info about the various job states.

The suggestion is to then use some settings closer those lower capacity setting when replicating to/from those endpoints (not sure which on of your endpoint is throttled, source, target or both).

**Comment by jdai1** [comment]
Thanks so much Nick! That is a big help, once again. It turns out the problem was the use_checkpoints. I am trying to clone the npm registry from https://replicate.npmjs.com/, which is a massive database, so I'm running into quite a few bugs. 
<img width="667" alt="Screenshot 2023-06-14 at 11 02 38 PM" src="https://github.com/apache/couchdb/assets/44037521/0970baf4-f966-4a52-9897-7ad9a4012003">
This is my current config for replication, which does not give me any checkpoint failures (I guess the couchdb for the npm registry was blocking any checkpoints). However, now, the replications crash with a new error!
```
{"error":"error","reason":"{worker_died,<0.5879.61>,{process_died,<0.7372.75>,kaboom}}"}
```
Since I disabled checkpointing, each time the replication task fails, it restarts, so as a result, the progress is super slow. I did see this thread[https://github.com/apache/couchdb/issues/745], and I was wondering if you guys have found a workaround for this bug.
My database currently looks like this:
```
"doc_del_count":165197,"doc_count":1565
```
and the actual database looks like this:
```
"doc_count":3279735,"doc_del_count":671697
```
So there is still a ways to go! What do you think would be best?

**Comment by smulikHakipod** [comment]
Did you guys found any soultion? I am trying to close replicate.npmjs.com with the same error. 
I tired using @nickva parameters without much success :(

**Comment by nickva** [comment]
@smulikHakipod the issue is that the source doesn't allow `_ensure_full_commit` so you'd have to try to replicate with `use_checkpoints = false` settings.


@jdai1 The replication is probably timing out or throttling our replication job. I'd suggest trying much lower settings like 1 worker and a batch of 100 or so only. With higher retries_per_request and a high timeout.

There is a hidden (undocumented) `"since_seq": "...-......"` replication job setting that allows starting a replication changes sequence from a particular start sequence (instead of 0) once you've replicated far enough you save the update sequence then start using that as the `"since_seq":...` in the replication job and so on.
